clang/test/Analysis/store-to-immutable-basic.cpp

```diff
@@ -0,0 +1,166 @@
+
+void test_write_to_const_global_struct() {
+  int *ptr = (int*)&global_struct.x;
+  *ptr = 10; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const parameter
+void test_write_to_const_param(const int param) { // expected-note {{Memory region is in immutable space}}
+  *(int*)&param = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const reference parameter
+void test_write_to_const_ref_param(const int &param) {
+  *(int*)&param = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const pointer parameter
+void test_write_to_const_ptr_param(const int *param) {
+  *(int*)param = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const array parameter
+void test_write_to_const_array_param(const int arr[5]) {
+  *(int*)arr = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const struct parameter
+struct ParamStruct {
+  const int z; // expected-note 3{{Memory region is in immutable space}}
+  int w;
+};
+
+void test_write_to_const_struct_param(const ParamStruct s) {
+  *(int*)&s.z = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const struct reference parameter
+void test_write_to_const_struct_ref_param(const ParamStruct &s) {
+  *(int*)&s.z = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const struct pointer parameter
+void test_write_to_const_struct_ptr_param(const ParamStruct *s) {
+  *(int*)&s->z = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+//===--- NEGATIVE TEST CASES ---===//
+// These tests should NOT trigger warnings
```

Comment:
This comment acts like a spooky action at a distance.
One would only see a small part of a PR in the future modifying this file and unknowingly break this comment.
I'd rather just not partition the tests into segments, because it's really difficult to enforce.

Reply:
I have removed this.